src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs

```diff
@@ -0,0 +1,229 @@
+            binding.ErrorContainer.MergeErrors(errors);
+
+            return binding;
+        }
+
+        private RecordType CheckParameters(out List<TexlError> errors)
+        {
+            var record = RecordType.Empty();
+            var argsAlreadySeen = new HashSet<string>();
+            errors = new List<TexlError>();
+
+            foreach (var arg in _args)
+            {
+                if (!argsAlreadySeen.Add(arg.VarIdent.Name))
+                {
+                    errors.Add(new TexlError(arg.VarIdent, DocumentErrorSeverity.Severe, TexlStrings.ErrUDF_DuplicateParameter, arg.VarIdent.Name));
+                }
+                else
+                {
+                    argsAlreadySeen.Add(arg.VarIdent.Name);
+                    record = record.Add(new NamedFormulaType(arg.VarIdent.ToString(), arg.VarType.GetFormulaType()));
+                }
+            }
+
+            return record;
+        }
+
+        /// <summary>
+        /// Perform sub-expression type checking and produce a return type for the function declaration, this is only applicable for UDFs.
+        /// </summary>
+        public static void CheckTypesOnDeclaration(CheckTypesContext context, IEnumerable<UDFArg> uDFArgs, IdentToken returnType, DType actualBodyReturnType, IErrorContainer errorContainer)
+        {
+            Contracts.AssertValue(context);
+            Contracts.AssertValue(uDFArgs);
+            Contracts.AssertValue(returnType);
+            Contracts.AssertValue(actualBodyReturnType);
+            Contracts.AssertValue(errorContainer);
+
+            CheckParameters(uDFArgs, errorContainer);
+            CheckReturnType(returnType, actualBodyReturnType, errorContainer);
+        }
+
+        private static void CheckReturnType(IdentToken returnType, DType actualBodyReturnType, IErrorContainer errorContainer)
+        {
+            var returnTypeFormulaType = returnType.GetFormulaType()._type;
+
+            if (returnTypeFormulaType.Kind.Equals(DType.Unknown.Kind))
+            {
+                errorContainer.EnsureError(DocumentErrorSeverity.Severe, returnType, TexlStrings.ErrUDF_UnknownType, returnType.Name);
+
+                // Anyhow the next error will be ignored, since it is added on the same token
+                return; 
+            }
+
+            if (!returnTypeFormulaType.Kind.Equals(actualBodyReturnType.Kind) || !returnTypeFormulaType.CoercesTo(returnTypeFormulaType))
+            {
+                errorContainer.EnsureError(DocumentErrorSeverity.Severe, returnType, TexlStrings.ErrUDF_ReturnTypeDoesNotMatch);
+            }
+        }
+
+        private static void CheckParameters(IEnumerable<UDFArg> args, IErrorContainer errorContainer)
+        {
+            foreach (var arg in args)
+            {
+                var argNameToken = arg.VarIdent;
+                var argTypeToken = arg.VarType;
+
+                if (argTypeToken.GetFormulaType()._type.Kind.Equals(DType.Unknown.Kind))
```

Comment:
We also need to guarantee that the parameter type is supported right? For example, if the parameter type is Void or some other unsupported type, we should check here. I know that the parser will also do a similar check, however we should be able to use this API in contexts that don't involve the Parser.

Reply:
Yup, this is done in CheckTypesOnDeclaration